cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -397,6 +397,136 @@ struct PowerChecked {
   }
 };
 
+// Stamp out a trig function for which +/-inf are domain errors
+#define TRIG_NO_INF(name, func)                                                        \
+  struct Trig##name {                                                                  \
+    template <typename T, typename Arg0>                                               \
+    static enable_if_integer<Arg0, T> Call(KernelContext*, Arg0 val, Status*) {        \
+      static_assert(std::is_same<T, double>::value, "");                               \
+      return func(val);                                                                \
+    }                                                                                  \
+    template <typename T, typename Arg0>                                               \
+    static enable_if_floating_point<Arg0, T> Call(KernelContext*, Arg0 val, Status*) { \
+      static_assert(std::is_same<T, Arg0>::value, "");                                 \
+      return func(val);                                                                \
+    }                                                                                  \
+  };                                                                                   \
+  struct Trig##name##Checked {                                                         \
+    template <typename T, typename Arg0>                                               \
+    static enable_if_integer<Arg0, T> Call(KernelContext*, Arg0 val, Status*) {        \
+      static_assert(std::is_same<T, double>::value, "");                               \
+      return func(val);                                                                \
+    }                                                                                  \
+    template <typename T, typename Arg0>                                               \
+    static enable_if_floating_point<Arg0, T> Call(KernelContext*, Arg0 val,            \
+                                                  Status* st) {                        \
+      static_assert(std::is_same<T, Arg0>::value, "");                                 \
+      if (ARROW_PREDICT_FALSE(std::isinf(val))) {                                      \
+        *st = Status::Invalid("domain error");                                         \
+        return val;                                                                    \
+      }                                                                                \
+      return func(val);                                                                \
+    }                                                                                  \
+  };
+
+TRIG_NO_INF(Sin, std::sin);
+TRIG_NO_INF(Cos, std::cos);
+TRIG_NO_INF(Tan, std::tan);
```

Comment:
I suggest to not use macros for defining compute functions and instead expand the definition of each function explicitly. I understand it is to reduce code duplication, but the same case could be applied to kernel registration code and others.